File: minimal/go/internal/core/providers/openai.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"

	"minimal-go/internal/config"
	"minimal-go/internal/types"
)

type openAIProvider struct {
	client  *http.Client
	apiKey  string
	baseURL string
}

type openAIMessage struct {
	Role       string           `json:"role"`
	Content    string           `json:"content,omitempty"`
	ToolCallID string           `json:"tool_call_id,omitempty"`
	ToolCalls  []openAIToolCall `json:"tool_calls,omitempty"`
}

type openAIToolCall struct {
	ID       string             `json:"id"`
	Type     string             `json:"type"`
	Function openAIToolFunction `json:"function"`
}

type openAIToolFunction struct {
	Name      string `json:"name"`
	Arguments string `json:"arguments"`
}

type openAITool struct {
	Type     string           `json:"type"`
	Function openAIToolSchema `json:"function"`
}

type openAIToolSchema struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	Parameters  map[string]interface{} `json:"parameters"`
}

type openAIRequest struct {
	Model       string          `json:"model"`
	Temperature float64         `json:"temperature"`
	MaxTokens   int             `json:"max_tokens"`
	Messages    []openAIMessage `json:"messages"`
	Tools       []openAITool    `json:"tools,omitempty"`
	ToolChoice  string          `json:"tool_choice,omitempty"`
}

type openAIResponse struct {
	Choices []struct {
		Message struct {
			Role      string           `json:"role"`
			Content   string           `json:"content"`
			ToolCalls []openAIToolCall `json:"tool_calls"`
		} `json:"message"`
	} `json:"choices"`
	Usage *struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
		TotalTokens      int `json:"total_tokens"`
	} `json:"usage"`
	Error *struct {
		Message string `json:"message"`
		Type    string `json:"type"`
		Code    string `json:"code"`
	} `json:"error"`
}

func NewOpenAIProvider(cfg config.ResolvedLlmConfig) ChatProvider {
	return &openAIProvider{
		client:  &http.Client{Timeout: 60 * time.Second},
		apiKey:  cfg.APIKey,
		baseURL: cfg.BaseURL,
	}
}

func (p *openAIProvider) CreateChatCompletion(params CreateChatParams) (ChatResponse, error) {
	requestParams := openAIRequest{
		Model:       params.Model,
		Temperature: params.Temperature,
		MaxTokens:   params.MaxTokens,
		Messages:    toOpenAIMessages(params.Messages),
		Tools:       toOpenAITools(params.Tools),
		ToolChoice:  "auto",
	}

	payload, err := json.Marshal(requestParams)
	if err != nil {
		return ChatResponse{}, err
	}

	endpoint, err := url.JoinPath(p.baseURL, "chat", "completions")
	if err != nil {
		return ChatResponse{}, err
	}

	req, err := http.NewRequest(http.MethodPost, endpoint, bytes.NewBuffer(payload))
	if err != nil {
		return ChatResponse{}, err
	}
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", p.apiKey))
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return ChatResponse{}, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return ChatResponse{}, err
	}

	var decoded openAIResponse
	if err := json.Unmarshal(body, &decoded); err != nil {
		return ChatResponse{}, err
	}
	if resp.StatusCode >= 300 {
		if decoded.Error != nil && decoded.Error.Message != "" {
			return ChatResponse{}, fmt.Errorf("%s (status %d)", decoded.Error.Message, resp.StatusCode)
		}
		return ChatResponse{}, fmt.Errorf("request failed with status %d", resp.StatusCode)
	}
	if decoded.Error != nil {
		return ChatResponse{}, fmt.Errorf(decoded.Error.Message)
	}

	message := types.Message{Role: types.RoleAssistant, Content: ""}
	var toolCalls []types.ToolCall
	if len(decoded.Choices) > 0 {
		choice := decoded.Choices[0]
		message.Content = choice.Message.Content
		for _, call := range choice.Message.ToolCalls {
			if call.Type != "function" {
				continue
			}
			toolCalls = append(toolCalls, types.ToolCall{
				ID:    call.ID,
				Name:  call.Function.Name,
				Input: parseToolInput(call.Function.Arguments),
			})
		}
	}
	if len(toolCalls) > 0 {
		message.ToolCalls = toolCalls
	}

	var usage *types.Usage
	if decoded.Usage != nil {
		usage = &types.Usage{
			PromptTokens:     decoded.Usage.PromptTokens,
			CompletionTokens: decoded.Usage.CompletionTokens,
			TotalTokens:      decoded.Usage.TotalTokens,
		}
	}

	return ChatResponse{
		Message:    message,
		Usage:      usage,
		RawUsage:   decoded.Usage,
		RawRequest: requestParams,
	}, nil
}

func toOpenAITools(tools []types.Tool) []openAITool {
	result := make([]openAITool, 0, len(tools))
	for _, tool := range tools {
		result = append(result, openAITool{
			Type: "function",
			Function: openAIToolSchema{
				Name:        tool.Name,
				Description: tool.Description,
				Parameters:  tool.InputSchema,
			},
		})
	}
	return result
}

func toOpenAIMessages(messages []types.Message) []openAIMessage {
	result := make([]openAIMessage, 0, len(messages))
	for _, message := range messages {
		switch message.Role {
		case types.RoleTool:
			result = append(result, openAIMessage{
				Role:       "tool",
				ToolCallID: message.ToolCallID,
				Content:    message.Content,
			})
		case types.RoleAssistant:
			if len(message.ToolCalls) > 0 {
				calls := make([]openAIToolCall, 0, len(message.ToolCalls))
				for _, call := range message.ToolCalls {
					args, _ := json.Marshal(call.Input)
					calls = append(calls, openAIToolCall{
						ID:   call.ID,
						Type: "function",
						Function: openAIToolFunction{
							Name:      call.Name,
							Arguments: string(args),
						},
					})
				}
				result = append(result, openAIMessage{
					Role:      "assistant",
					Content:   message.Content,
					ToolCalls: calls,
				})
			} else {
				result = append(result, openAIMessage{
					Role:    "assistant",
					Content: message.Content,
				})
			}
		default:
			result = append(result, openAIMessage{
				Role:    string(message.Role),
				Content: message.Content,
			})
		}
	}
	return result
}

func parseToolInput(raw string) interface{} {
	if raw == "" {
		return map[string]interface{}{}
	}
	var parsed interface{}
	if err := json.Unmarshal([]byte(raw), &parsed); err != nil {
		return raw
	}
	return parsed
}

File: minimal/rust/src/core/providers/openai.rs
use crate::config::ResolvedLlmConfig;
use crate::types::{Message, Role, Tool, ToolCall, Usage};
use crate::core::providers::{ChatProvider, ChatResponse, CreateChatParams};
use reqwest::blocking::Client;
use reqwest::Url;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::time::Duration;

pub struct OpenAIProvider {
    client: Client,
    api_key: Option<String>,
    base_url: String,
}

#[derive(Serialize)]
struct OpenAIMessage {
    role: String,
    content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    tool_call_id: Option<String>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    tool_calls: Vec<OpenAIToolCall>,
}

#[derive(Serialize, Deserialize)]
struct OpenAIToolCall {
    id: String,
    #[serde(rename = "type")]
    call_type: String,
    function: OpenAIToolFunction,
}

#[derive(Serialize, Deserialize)]
struct OpenAIToolFunction {
    name: String,
    arguments: String,
}

#[derive(Serialize)]
struct OpenAITool {
    #[serde(rename = "type")]
    tool_type: String,
    function: OpenAIToolSchema,
}

#[derive(Serialize)]
struct OpenAIToolSchema {
    name: String,
    description: String,
    parameters: Value,
}

#[derive(Serialize)]
struct OpenAIRequest {
    model: String,
    temperature: f64,
    max_tokens: i32,
    messages: Vec<OpenAIMessage>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    tools: Vec<OpenAITool>,
    tool_choice: String,
}

#[derive(Deserialize)]
struct OpenAIResponse {
    choices: Vec<OpenAIChoice>,
    usage: Option<OpenAIUsage>,
    error: Option<OpenAIError>,
}

#[derive(Deserialize)]
struct OpenAIChoice {
    message: OpenAIResponseMessage,
}

#[derive(Deserialize)]
struct OpenAIResponseMessage {
    role: String,
    content: Option<String>,
    #[serde(default)]
    tool_calls: Vec<OpenAIToolCall>,
}

#[derive(Deserialize, Serialize)]
struct OpenAIUsage {
    prompt_tokens: i32,
    completion_tokens: i32,
    total_tokens: i32,
}

#[derive(Deserialize)]
struct OpenAIError {
    message: Option<String>,
}

impl OpenAIProvider {
    pub fn new(cfg: &ResolvedLlmConfig) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(60))
            .build()
            .expect("Failed to build HTTP client");

        Self {
            client,
            api_key: cfg.api_key.clone(),
            base_url: cfg.base_url.clone(),
        }
    }
}

impl ChatProvider for OpenAIProvider {
    fn create_chat_completion(&self, params: CreateChatParams) -> Result<ChatResponse, String> {
        let request_params = OpenAIRequest {
            model: params.model.clone(),
            temperature: params.temperature,
            max_tokens: params.max_tokens,
            messages: to_openai_messages(&params.messages),
            tools: to_openai_tools(&params.tools),
            tool_choice: "auto".to_string(),
        };

        let payload = serde_json::to_vec(&request_params)
            .map_err(|err| format!("Failed to encode request: {}", err))?;

        let endpoint = Url::parse(&self.base_url)
            .and_then(|base| base.join("chat/completions"))
            .map_err(|err| format!("Invalid base_url: {}", err))?;

        let mut req = self
            .client
            .post(endpoint)
            .header("Content-Type", "application/json");

        if let Some(api_key) = &self.api_key {
            req = req.header("Authorization", format!("Bearer {}", api_key));
        }

        let resp = req
            .body(payload)
            .send()
            .map_err(|err| format!("Request failed: {}", err))?;

        let status = resp.status();
        let body = resp
            .text()
            .map_err(|err| format!("Failed to read response: {}", err))?;

        let decoded: OpenAIResponse = serde_json::from_str(&body)
            .map_err(|err| format!("Invalid response: {}", err))?;

        if !status.is_success() {
            if let Some(err) = decoded.error.and_then(|e| e.message) {
                return Err(format!("{} (status {})", err, status.as_u16()));
            }
            return Err(format!("request failed with status {}", status.as_u16()));
        }

        if let Some(err) = decoded.error.and_then(|e| e.message) {
            return Err(err);
        }

        let mut message = Message {
            role: Role::Assistant,
            content: String::new(),
            thinking: None,
            tool_call_id: None,
            tool_calls: Vec::new(),
        };

        let mut tool_calls: Vec<ToolCall> = Vec::new();
        if let Some(choice) = decoded.choices.first() {
            message.content = choice.message.content.clone().unwrap_or_default();
            for call in &choice.message.tool_calls {
                if call.call_type != "function" {
                    continue;
                }
                tool_calls.push(ToolCall {
                    id: call.id.clone(),
                    name: call.function.name.clone(),
                    input: parse_tool_input(&call.function.arguments),
                });
            }
        }

        if !tool_calls.is_empty() {
            message.tool_calls = tool_calls;
        }

        let (usage, raw_usage) = match decoded.usage {
            Some(usage) => {
                let raw = serde_json::to_value(&usage).unwrap_or(json!({}));
                (
                    Some(Usage {
                        prompt_tokens: usage.prompt_tokens,
                        completion_tokens: usage.completion_tokens,
                        total_tokens: usage.total_tokens,
                    }),
                    Some(raw),
                )
            }
            None => (None, None),
        };

        Ok(ChatResponse {
            message,
            usage,
            raw_usage,
            raw_request: serde_json::to_value(&request_params).unwrap_or(json!({})),
            raw_headers: None,
        })
    }
}

fn to_openai_tools(tools: &[Tool]) -> Vec<OpenAITool> {
    tools
        .iter()
        .map(|tool| OpenAITool {
            tool_type: "function".to_string(),
            function: OpenAIToolSchema {
                name: tool.name.clone(),
                description: tool.description.clone(),
                parameters: tool.input_schema.clone(),
            },
        })
        .collect()
}

fn to_openai_messages(messages: &[Message]) -> Vec<OpenAIMessage> {
    let mut result = Vec::new();
    for message in messages {
        match message.role {
            Role::Tool => {
                result.push(OpenAIMessage {
                    role: "tool".to_string(),
                    content: message.content.clone(),
                    tool_call_id: message.tool_call_id.clone(),
                    tool_calls: Vec::new(),
                });
            }
            Role::Assistant => {
                if !message.tool_calls.is_empty() {
                    let mut calls = Vec::new();
                    for call in &message.tool_calls {
                        let args = serde_json::to_string(&call.input).unwrap_or("{}".to_string());
                        calls.push(OpenAIToolCall {
                            id: call.id.clone(),
                            call_type: "function".to_string(),
                            function: OpenAIToolFunction {
                                name: call.name.clone(),
                                arguments: args,
                            },
                        });
                    }
                    result.push(OpenAIMessage {
                        role: "assistant".to_string(),
                        content: message.content.clone(),
                        tool_call_id: None,
                        tool_calls: calls,
                    });
                } else {
                    result.push(OpenAIMessage {
                        role: "assistant".to_string(),
                        content: message.content.clone(),
                        tool_call_id: None,
                        tool_calls: Vec::new(),
                    });
                }
            }
            _ => {
                result.push(OpenAIMessage {
                    role: match message.role {
                        Role::System => "system".to_string(),
                        Role::User => "user".to_string(),
                        Role::Assistant => "assistant".to_string(),
                        Role::Tool => "tool".to_string(),
                    },
                    content: message.content.clone(),
                    tool_call_id: None,
                    tool_calls: Vec::new(),
                });
            }
        }
    }

    result
}

fn parse_tool_input(raw: &str) -> Value {
    if raw.is_empty() {
        return json!({});
    }
    serde_json::from_str(raw).unwrap_or_else(|_| Value::String(raw.to_string()))
}
